File: lib/services/smart_categorization_service.dart
class SmartCategorizationService {
  static SmartCategorizationService? _instance;
  static SmartCategorizationService get instance =>
      _instance ??= SmartCategorizationService._();
  SmartCategorizationService._();

  // Priority determination based on keywords and category
  String determinePriority(String title, String description, String category) {
    final text = '${title.toLowerCase()} ${description.toLowerCase()}';

    // Critical priority keywords
    final criticalKeywords = [
      'emergency',
      'urgent',
      'dangerous',
      'hazardous',
      'life threatening',
      'fire',
      'gas leak',
      'water main burst',
      'electrical hazard',
      'accident',
      'injury',
      'blocked ambulance',
      'major flooding',
      'bridge collapse',
      'building collapse',
      'explosion',
      'chemical spill',
    ];

    // High priority keywords
    final highKeywords = [
      'serious',
      'major',
      'significant',
      'important',
      'broken main',
      'power outage',
      'sewage overflow',
      'traffic light broken',
      'road blocked',
      'sinkhole',
      'falling debris',
      'unsafe conditions',
      'broken water main',
      'severe pothole',
      'damaged bridge',
    ];

    // Medium priority keywords
    final mediumKeywords = [
      'moderate',
      'noticeable',
      'concerning',
      'needs attention',
      'street light out',
      'small pothole',
      'trash overflow',
      'graffiti',
      'noise complaint',
      'minor leak',
      'cracked pavement',
    ];

    // Check for critical indicators
    if (_containsAny(text, criticalKeywords) || _isCriticalCategory(category)) {
      return 'Critical';
    }

    // Check for high priority indicators
    if (_containsAny(text, highKeywords) || _isHighPriorityCategory(category)) {
      return 'High';
    }

    // Check for medium priority indicators
    if (_containsAny(text, mediumKeywords)) {
      return 'Medium';
    }

    // Default priority based on category
    return _getDefaultPriorityForCategory(category);
  }

  // Enhanced department assignment based on intelligent keyword analysis
  String determineDepartment(
    String title,
    String description,
    String category,
  ) {
    final text = '${title.toLowerCase()} ${description.toLowerCase()}';

    // Emergency services - highest priority overrides
    if (_containsAny(text, [
      'fire',
      'smoke',
      'burning',
      'explosion',
      'flames',
      'gas leak',
    ])) {
      return 'Fire Department';
    }

    if (_containsAny(text, [
      'crime',
      'theft',
      'robbery',
      'vandalism',
      'suspicious',
      'violence',
      'break-in',
      'assault',
      'harassment',
      'drug',
      'illegal',
    ])) {
      return 'Police Department';
    }

    if (_containsAny(text, [
      'medical emergency',
      'injury',
      'accident',
      'ambulance',
      'hurt',
      'bleeding',
      'unconscious',
      'heart attack',
      'emergency medical',
    ])) {
      return 'Emergency Medical Services';
    }

    // Water-related issues with specific department routing
    if (_containsAny(text, [
      'water',
      'pipe',
      'leak',
      'burst',
      'flooding',
      'drainage',
      'sewer',
      'overflow',
      'blockage',
      'tap',
      'valve',
      'water supply',
      'no water',
    ])) {
      return 'Water & Sewerage';
    }

    // Electrical issues
    if (_containsAny(text, [
      'electricity',
      'power',
      'electric',
      'voltage',
      'wire',
      'cable',
      'outage',
      'blackout',
      'transformer',
      'pole',
      'meter',
    ])) {
      return 'Electricity';
    }

    // Road and transportation
    if (_containsAny(text, [
      'road',
      'street',
      'highway',
      'traffic',
      'pothole',
      'pavement',
      'crossing',
      'bridge',
      'tunnel',
      'signal',
      'sign',
    ])) {
      if (_containsAny(text, [
        'traffic light',
        'signal',
        'stop sign',
        'crossing light',
      ])) {
        return 'Traffic Police';
      }
      return 'Transportation';
    }

    // Waste management and sanitation
    if (_containsAny(text, [
      'garbage',
      'trash',
      'waste',
      'rubbish',
      'bin',
      'collection',
      'dump',
      'litter',
      'cleaning',
      'sanitation',
    ])) {
      return 'Public Works';
    }

    // Parks and environment
    if (_containsAny(text, [
      'park',
      'garden',
      'tree',
      'grass',
      'playground',
      'bench',
      'landscaping',
      'plants',
      'flowers',
      'recreation',
    ])) {
      return 'Public Works';
    }

    // Health and environment
    if (_containsAny(text, [
      'health',
      'disease',
      'epidemic',
      'vaccination',
      'clinic',
      'hospital',
      'medical facility',
      'public health',
    ])) {
      return 'Health Services';
    }

    if (_containsAny(text, [
      'pollution',
      'environment',
      'air quality',
      'noise',
      'dust',
      'industrial waste',
      'smoke emission',
    ])) {
      return 'Environmental Services';
    }

    // Building and construction
    if (_containsAny(text, [
      'building',
      'construction',
      'structure',
      'illegal construction',
      'permit',
      'violation',
      'unauthorized',
    ])) {
      return 'Municipal Corporation';
    }

    // Animal control
    if (_containsAny(text, [
      'animal',
      'dog',
      'cat',
      'stray',
      'pet',
      'wildlife',
      'bird',
      'pest',
    ])) {
      return 'Social Services';
    }

    // Food safety and markets
    if (_containsAny(text, [
      'food',
      'restaurant',
      'market',
      'vendor',
      'food safety',
      'hygiene',
      'expired food',
    ])) {
      return 'Health Services';
    }

    // Education related
    if (_containsAny(text, [
      'school',
      'education',
      'teacher',
      'student',
      'classroom',
      'college',
    ])) {
      return 'Education';
    }

    // Transport and vehicles
    if (_containsAny(text, [
      'bus',
      'auto',
      'taxi',
      'vehicle',
      'transport',
      'metro',
      'railway',
    ])) {
      return 'Transportation';
    }

    // Category-based fallback assignment
    switch (category.toLowerCase()) {
      case 'road & transportation':
        return 'Transportation';
      case 'water & sewerage':
        return 'Water & Sewerage';
      case 'electricity':
        return 'Electricity';
      case 'garbage & sanitation':
        return 'Public Works';
      case 'street lighting':
        return 'Electricity';
      case 'parks & recreation':
        return 'Public Works';
      case 'public safety':
        return 'Traffic Police';
      case 'noise pollution':
        return 'Environmental Services';
      case 'health':
        return 'Health Services';
      case 'education':
        return 'Education';
      case 'other':
        return _determineOtherDepartment(text);
      default:
        return 'Municipal Corporation';
    }
  }

  // Get estimated resolution time based on priority and department
  String getEstimatedResolutionTime(String priority, String department) {
    final baseDays = _getBaseDaysForDepartment(department);

    switch (priority) {
      case 'Critical':
        return 'Within 24 hours';
      case 'High':
        return baseDays <= 2 ? 'Within $baseDays days' : 'Within 2-3 days';
      case 'Medium':
        return 'Within $baseDays days';
      case 'Low':
        return 'Within ${baseDays + 2}-${baseDays + 7} days';
      default:
        return 'Within $baseDays days';
    }
  }

  // Get department contact information
  Map<String, String> getDepartmentContact(String department) {
    final contacts = {
      'Fire Department': {
        'phone': '911',
        'email': 'fire.emergency@city.gov',
        'hours': '24/7',
      },
      'Police Department': {
        'phone': '911',
        'email': 'police@city.gov',
        'hours': '24/7',
      },
      'Emergency Medical Services': {
        'phone': '911',
        'email': 'emergency@city.gov',
        'hours': '24/7',
      },
      'Public Works - Roads': {
        'phone': '(555) 123-4567',
        'email': 'roads@city.gov',
        'hours': 'Mon-Fri 8AM-5PM',
      },
      'Water Department': {
        'phone': '(555) 123-4568',
        'email': 'water@city.gov',
        'hours': 'Mon-Fri 8AM-5PM, Emergency 24/7',
      },
      'Electric Utility': {
        'phone': '(555) 123-4569',
        'email': 'electric@city.gov',
        'hours': 'Mon-Fri 8AM-5PM, Emergency 24/7',
      },
      'Sanitation Department': {
        'phone': '(555) 123-4570',
        'email': 'sanitation@city.gov',
        'hours': 'Mon-Fri 6AM-4PM',
      },
    };

    return contacts[department] ??
        {
          'phone': '(555) 123-4500',
          'email': 'general@city.gov',
          'hours': 'Mon-Fri 8AM-5PM',
        };
  }

  // Comprehensive classification for real-world scenarios
  Map<String, String> classifyIssue(
    String title,
    String description,
    String category,
  ) {
    final priority = determinePriority(title, description, category);
    final department = determineDepartment(title, description, category);
    final estimatedTime = getEstimatedResolutionTime(priority, department);
    final contact = getDepartmentContact(department);

    return {
      'priority': priority,
      'department': department,
      'estimatedTime': estimatedTime,
      'contactPhone': contact['phone'] ?? '',
      'contactEmail': contact['email'] ?? '',
      'workingHours': contact['hours'] ?? '',
      'urgencyLevel': _getUrgencyLevel(priority),
      'actionRequired': _getActionRequired(priority, department),
    };
  }

  // Auto-classify issue based on description (main feature)
  Map<String, dynamic> classifyIssueStatement(
    String title,
    String description,
  ) {
    // Auto-detect category based on content
    String category = autoDetectCategory(title, description);

    // Classify department based on content
    String department = determineDepartment(title, description, category);

    // Get priority
    String priority = determinePriority(title, description, category);

    // Get estimated time
    String estimatedTime = getEstimatedResolutionTime(priority, department);

    // Get contact info
    Map<String, String> contactInfo = getDepartmentContact(department);

    return {
      'originalTitle': title,
      'originalDescription': description,
      'detectedCategory': category,
      'assignedDepartment': department,
      'priority': priority,
      'estimatedResolution': estimatedTime,
      'urgencyLevel': _getUrgencyLevel(priority),
      'actionRequired': _getActionRequired(priority, department),
      'contactPhone': contactInfo['phone'],
      'contactEmail': contactInfo['email'],
      'workingHours': contactInfo['hours'],
    };
  }

  // Private helper methods
  bool _containsAny(String text, List<String> keywords) {
    return keywords.any((keyword) => text.contains(keyword.toLowerCase()));
  }

  bool _isCriticalCategory(String category) {
    final criticalCategories = ['public safety'];
    return criticalCategories.contains(category.toLowerCase());
  }

  bool _isHighPriorityCategory(String category) {
    final highPriorityCategories = ['water & sewerage', 'electricity'];
    return highPriorityCategories.contains(category.toLowerCase());
  }

  String _getDefaultPriorityForCategory(String category) {
    switch (category.toLowerCase()) {
      case 'public safety':
      case 'water & sewerage':
      case 'electricity':
        return 'High';
      case 'road & transportation':
      case 'garbage & sanitation':
        return 'Medium';
      case 'street lighting':
      case 'parks & recreation':
      case 'noise pollution':
        return 'Low';
      default:
        return 'Medium';
    }
  }

  String _getUrgencyLevel(String priority) {
    switch (priority) {
      case 'Critical':
        return 'IMMEDIATE ACTION REQUIRED';
      case 'High':
        return 'Urgent - High Priority';
      case 'Medium':
        return 'Standard Priority';
      case 'Low':
        return 'Low Priority - Routine';
      default:
        return 'Standard Priority';
    }
  }

  String _getActionRequired(String priority, String department) {
    if (priority == 'Critical') {
      if (department.contains('Fire') ||
          department.contains('Police') ||
          department.contains('Emergency')) {
        return 'Emergency response team dispatched immediately';
      }
      return 'Priority dispatch within 2-4 hours';
    } else if (priority == 'High') {
      return 'Same day response expected';
    } else if (priority == 'Medium') {
      return 'Response within 2-3 business days';
    } else {
      return 'Response within 5-7 business days';
    }
  }

  String _determineOtherDepartment(String text) {
    // Advanced keyword analysis for uncategorized issues
    if (_containsAny(text, [
      'animal',
      'pet',
      'dog',
      'cat',
      'wildlife',
      'stray',
      'rabid',
      'bite',
    ])) {
      return 'Social Services';
    }

    if (_containsAny(text, [
      'building',
      'construction',
      'permit',
      'code violation',
      'illegal structure',
      'unauthorized building',
      'unsafe building',
      'demolition',
    ])) {
      return 'Municipal Corporation';
    }

    if (_containsAny(text, [
      'business',
      'license',
      'permit',
      'zoning',
      'commercial',
      'shop',
      'vendor',
      'market',
      'hawker',
    ])) {
      return 'Municipal Corporation';
    }

    if (_containsAny(text, [
      'health',
      'food safety',
      'restaurant',
      'contamination',
      'hygiene',
      'disease',
      'epidemic',
      'vaccination',
    ])) {
      return 'Health Services';
    }

    if (_containsAny(text, [
      'cemetery',
      'burial',
      'grave',
      'funeral',
      'cremation',
    ])) {
      return 'Social Services';
    }

    if (_containsAny(text, [
      'library',
      'book',
      'reading',
      'study hall',
      'community center',
    ])) {
      return 'Education';
    }

    if (_containsAny(text, [
      'social',
      'welfare',
      'pension',
      'disability',
      'elderly',
      'senior citizen',
    ])) {
      return 'Social Services';
    }

    if (_containsAny(text, [
      'tax',
      'property tax',
      'assessment',
      'revenue',
      'collection',
    ])) {
      return 'Municipal Corporation';
    }

    if (_containsAny(text, [
      'document',
      'certificate',
      'birth certificate',
      'death certificate',
      'marriage certificate',
      'records',
    ])) {
      return 'Municipal Corporation';
    }

    if (_containsAny(text, [
      'employment',
      'job',
      'unemployment',
      'skill development',
      'training',
    ])) {
      return 'Social Services';
    }

    if (_containsAny(text, [
      'fire safety',
      'fire clearance',
      'fire NOC',
      'fire certificate',
    ])) {
      return 'Public Works';
    }

    if (_containsAny(text, [
      'legal',
      'court',
      'lawyer',
      'advocate',
      'legal aid',
      'litigation',
    ])) {
      return 'Municipal Corporation';
    }

    if (_containsAny(text, [
      'women',
      'child',
      'domestic violence',
      'harassment',
      'protection',
    ])) {
      return 'Social Services';
    }

    if (_containsAny(text, [
      'tourism',
      'tourist',
      'heritage',
      'monument',
      'cultural',
    ])) {
      return 'Social Services';
    }

    return 'Municipal Corporation';
  }

  int _getBaseDaysForDepartment(String department) {
    switch (department) {
      case 'Fire Department':
      case 'Police Department':
      case 'Emergency Medical Services':
        return 1;
      case 'Public Works - Roads':
      case 'Water Department':
      case 'Electric Utility':
        return 3;
      case 'Sanitation Department':
      case 'Street Lighting Department':
        return 5;
      case 'Parks & Recreation':
      case 'Environmental Department':
        return 7;
      default:
        return 5;
    }
  }

  String autoDetectCategory(String title, String description) {
    final text = '${title.toLowerCase()} ${description.toLowerCase()}';

    if (_containsAny(text, [
      'water',
      'pipe',
      'leak',
      'sewer',
      'drainage',
      'flooding',
    ])) {
      return 'Water & Sewerage';
    }
    if (_containsAny(text, [
      'road',
      'street',
      'pothole',
      'traffic',
      'bridge',
      'highway',
    ])) {
      return 'Road & Transportation';
    }
    if (_containsAny(text, [
      'electricity',
      'power',
      'outage',
      'electric',
      'voltage',
    ])) {
      return 'Electricity';
    }
    if (_containsAny(text, [
      'garbage',
      'trash',
      'waste',
      'sanitation',
      'cleaning',
    ])) {
      return 'Garbage & Sanitation';
    }
    if (_containsAny(text, ['street light', 'lamp', 'lighting', 'bulb'])) {
      return 'Street Lighting';
    }
    if (_containsAny(text, [
      'park',
      'garden',
      'playground',
      'recreation',
      'tree',
    ])) {
      return 'Parks & Recreation';
    }
    if (_containsAny(text, [
      'crime',
      'safety',
      'police',
      'suspicious',
      'theft',
    ])) {
      return 'Public Safety';
    }
    if (_containsAny(text, ['noise', 'pollution', 'loud', 'sound'])) {
      return 'Noise Pollution';
    }
    if (_containsAny(text, ['health', 'medical', 'disease', 'food safety'])) {
      return 'Health';
    }
    if (_containsAny(text, ['education', 'school', 'teacher', 'student'])) {
      return 'Education';
    }

    return 'Other';
  }
}

File: backend/services/smartCategorizationService.js
// Smart Categorization Service
// This service provides AI-powered categorization for reports

const categoryKeywords = {
  garbage_collection: [
    'garbage', 'waste', 'trash', 'rubbish', 'dump', 'dustbin', 'litter', 
    'disposal', 'collection', 'cleaning', 'sweeping', 'sanitation',
    'dirty', 'smell', 'odor', 'flies', 'rats', 'pest'
  ],
  
  road_maintenance: [
    'road', 'street', 'pothole', 'crack', 'broken', 'damaged', 'repair',
    'asphalt', 'concrete', 'pavement', 'footpath', 'sidewalk', 'pathway',
    'surface', 'construction', 'maintenance', 'fix'
  ],
  
  street_lights: [
    'light', 'lamp', 'electricity', 'power', 'dark', 'bright', 'bulb',
    'illumination', 'lighting', 'street light', 'pole', 'electric',
    'night', 'visibility', 'safety', 'broken light', 'not working'
  ],
  
  water_supply: [
    'water', 'tap', 'pipe', 'pipeline', 'leak', 'burst', 'supply',
    'pressure', 'flow', 'shortage', 'contaminated', 'dirty water',
    'no water', 'valve', 'connection', 'plumbing', 'drinking water'
  ],
  
  drainage: [
    'drain', 'sewer', 'sewage', 'overflow', 'clog', 'blocked', 'flood',
    'water logging', 'stagnant', 'manhole', 'gutter', 'canal',
    'waterlogged', 'rainwater', 'storm drain', 'blockage'
  ],
  
  public_safety: [
    'safety', 'security', 'crime', 'theft', 'robbery', 'violence',
    'danger', 'accident', 'emergency', 'police', 'patrol', 'guard',
    'protection', 'harassment', 'unsafe', 'criminal', 'incident'
  ],
  
  health_services: [
    'health', 'hospital', 'clinic', 'medical', 'doctor', 'nurse',
    'medicine', 'treatment', 'disease', 'illness', 'epidemic',
    'vaccination', 'ambulance', 'emergency', 'patient', 'healthcare'
  ],
  
  transportation: [
    'bus', 'auto', 'rickshaw', 'taxi', 'vehicle', 'transport',
    'traffic', 'signal', 'congestion', 'parking', 'stand',
    'stop', 'route', 'service', 'public transport', 'metro'
  ],
  
  noise_pollution: [
    'noise', 'loud', 'sound', 'music', 'speaker', 'horn', 'construction',
    'disturbance', 'pollution', 'party', 'festival', 'generator',
    'machine', 'vehicle noise', 'acoustic', 'decibel'
  ],
  
  air_pollution: [
    'air', 'pollution', 'smoke', 'dust', 'emission', 'smog', 'factory',
    'exhaust', 'fumes', 'burning', 'industrial', 'vehicle emission',
    'quality', 'breathing', 'asthma', 'cough'
  ],
  
  illegal_construction: [
    'illegal', 'construction', 'building', 'unauthorized', 'permit',
    'violation', 'encroachment', 'structure', 'demolition', 'planning',
    'approval', 'land', 'property', 'development'
  ],
  
  traffic_management: [
    'traffic', 'signal', 'jam', 'congestion', 'rule', 'violation',
    'parking', 'zebra crossing', 'speed', 'lane', 'intersection',
    'management', 'control', 'flow', 'vehicle'
  ],
  
  public_facilities: [
    'park', 'garden', 'playground', 'toilet', 'public', 'facility',
    'maintenance', 'bench', 'shelter', 'community', 'center',
    'library', 'hall', 'ground', 'sports'
  ]
};

const priorityKeywords = {
  critical: [
    'emergency', 'urgent', 'immediate', 'critical', 'danger', 'life',
    'death', 'fire', 'explosion', 'collapse', 'accident', 'injury',
    'bleeding', 'severe', 'crisis', 'disaster'
  ],
  
  high: [
    'important', 'serious', 'major', 'significant', 'broken', 'damaged',
    'leaking', 'flooding', 'blocked', 'safety', 'health', 'risk'
  ],
  
  medium: [
    'issue', 'problem', 'concern', 'maintenance', 'repair', 'fix',
    'improve', 'update', 'clean', 'service'
  ],
  
  low: [
    'suggestion', 'recommendation', 'minor', 'cosmetic', 'aesthetic',
    'enhancement', 'upgrade', 'convenience'
  ]
};

const urgencyKeywords = {
  emergency: [
    'emergency', 'immediate', 'now', 'asap', 'urgent', 'critical',
    'life threatening', 'danger', 'crisis', 'disaster'
  ],
  
  high: [
    'urgent', 'soon', 'priority', 'important', 'quickly', 'fast',
    'immediate attention', 'serious'
  ],
  
  medium: [
    'moderate', 'normal', 'regular', 'standard', 'typical', 'usual'
  ],
  
  low: [
    'whenever', 'eventually', 'later', 'convenience', 'minor',
    'not urgent', 'low priority'
  ]
};

class SmartCategorizationService {
  
  /**
   * Categorize a report based on title, description, and other factors
   * @param {Object} reportData - The report data to categorize
   * @returns {Object} - Categorization result with confidence score
   */
  async categorizeReport(reportData) {
    try {
      const { title, description, category, location } = reportData;
      
      // Combine title and description for analysis
      const text = `${title} ${description}`.toLowerCase();
      
      // Calculate category confidence
      const categoryResult = this.categorizeByKeywords(text, categoryKeywords);
      
      // Calculate priority
      const priorityResult = this.categorizeByKeywords(text, priorityKeywords);
      
      // Calculate urgency
      const urgencyResult = this.categorizeByKeywords(text, urgencyKeywords);
      
      // Extract keywords
      const extractedKeywords = this.extractKeywords(text);
      
      // Location-based adjustments
      const locationAdjustments = this.getLocationAdjustments(location);
      
      // Time-based adjustments (e.g., night-time reports might be more urgent)
      const timeAdjustments = this.getTimeAdjustments();
      
      return {
        suggestedCategory: categoryResult.category || category,
        confidence: categoryResult.confidence,
        suggestedPriority: priorityResult.category || 'medium',
        priorityConfidence: priorityResult.confidence,
        suggestedUrgency: urgencyResult.category || 'medium',
        urgencyConfidence: urgencyResult.confidence,
        keywords: extractedKeywords,
        locationFactors: locationAdjustments,
        timeFactors: timeAdjustments,
        recommendations: this.generateRecommendations(categoryResult, priorityResult, urgencyResult),
        estimatedResolutionTime: this.estimateResolutionTime(categoryResult.category, priorityResult.category)
      };
      
    } catch (error) {
      console.error('Smart categorization error:', error);
      return {
        suggestedCategory: reportData.category || 'others',
        confidence: 0,
        suggestedPriority: 'medium',
        priorityConfidence: 0,
        suggestedUrgency: 'medium',
        urgencyConfidence: 0,
        keywords: [],
        error: error.message
      };
    }
  }
  
  /**
   * Categorize text based on keyword matching
   * @param {string} text - Text to analyze
   * @param {Object} keywords - Category keywords object
   * @returns {Object} - Category and confidence score
   */
  categorizeByKeywords(text, keywords) {
    const scores = {};
    let totalMatches = 0;
    
    // Calculate scores for each category
    Object.keys(keywords).forEach(category => {
      let categoryScore = 0;
      let matches = 0;
      
      keywords[category].forEach(keyword => {
        const regex = new RegExp(`\\b${keyword}\\b`, 'gi');
        const keywordMatches = (text.match(regex) || []).length;
        
        if (keywordMatches > 0) {
          matches++;
          // Weight longer keywords more heavily
          categoryScore += keywordMatches * (keyword.length > 8 ? 2 : 1);
        }
      });
      
      scores[category] = {
        score: categoryScore,
        matches: matches,
        percentage: matches / keywords[category].length
      };
      
      totalMatches += categoryScore;
    });
    
    // Find the category with highest score
    let bestCategory = null;
    let bestScore = 0;
    
    Object.keys(scores).forEach(category => {
      if (scores[category].score > bestScore) {
        bestScore = scores[category].score;
        bestCategory = category;
      }
    });
    
    // Calculate confidence (0-1 scale)
    const confidence = totalMatches > 0 ? Math.min(bestScore / totalMatches, 1) : 0;
    
    return {
      category: bestCategory,
      confidence: confidence,
      scores: scores,
      totalMatches: totalMatches
    };
  }
  
  /**
   * Extract relevant keywords from text
   * @param {string} text - Text to extract keywords from
   * @returns {Array} - Array of extracted keywords
   */
  extractKeywords(text) {
    const stopWords = new Set([
      'the', 'a', 'an', 'and', 'or', 'but', 'in', 'on', 'at', 'to', 'for', 
      'of', 'with', 'by', 'is', 'are', 'was', 'were', 'be', 'been', 'have', 
      'has', 'had', 'do', 'does', 'did', 'will', 'would', 'could', 'should',
      'there', 'here', 'this', 'that', 'these', 'those', 'i', 'you', 'he', 
      'she', 'it', 'we', 'they', 'my', 'your', 'his', 'her', 'its', 'our', 'their'
    ]);
    
    const words = text.toLowerCase()
      .replace(/[^\w\s]/g, ' ')
      .split(/\s+/)
      .filter(word => word.length > 2 && !stopWords.has(word));
    
    // Count word frequencies
    const wordCount = {};
    words.forEach(word => {
      wordCount[word] = (wordCount[word] || 0) + 1;
    });
    
    // Return top keywords sorted by frequency
    return Object.keys(wordCount)
      .sort((a, b) => wordCount[b] - wordCount[a])
      .slice(0, 10); // Top 10 keywords
  }
  
  /**
   * Get location-based adjustments
   * @param {Object} location - Location data
   * @returns {Object} - Location adjustment factors
   */
  getLocationAdjustments(location) {
    const adjustments = {
      urbanBonus: 0,
      densityFactor: 1,
      accessibilityFactor: 1
    };
    
    if (location) {
      // Urban areas might have different priority weights
      const urbanKeywords = ['city', 'metro', 'urban', 'downtown', 'center'];
      const locationText = `${location.address || ''} ${location.city || ''}`.toLowerCase();
      
      if (urbanKeywords.some(keyword => locationText.includes(keyword))) {
        adjustments.urbanBonus = 0.1;
        adjustments.densityFactor = 1.2;
      }
      
      // Remote areas might need higher urgency
      const remoteKeywords = ['village', 'rural', 'remote', 'outskirts'];
      if (remoteKeywords.some(keyword => locationText.includes(keyword))) {
        adjustments.accessibilityFactor = 1.3;
      }
    }
    
    return adjustments;
  }
  
  /**
   * Get time-based adjustments
   * @returns {Object} - Time adjustment factors
   */
  getTimeAdjustments() {
    const now = new Date();
    const hour = now.getHours();
    const isWeekend = now.getDay() === 0 || now.getDay() === 6;
    
    return {
      isOfficeHours: hour >= 9 && hour <= 17,
      isNightTime: hour >= 22 || hour <= 6,
      isWeekend: isWeekend,
      urgencyMultiplier: (hour >= 22 || hour <= 6) ? 1.2 : 1.0
    };
  }
  
  /**
   * Generate recommendations based on categorization results
   * @param {Object} categoryResult - Category analysis result
   * @param {Object} priorityResult - Priority analysis result
   * @param {Object} urgencyResult - Urgency analysis result
   * @returns {Array} - Array of recommendations
   */
  generateRecommendations(categoryResult, priorityResult, urgencyResult) {
    const recommendations = [];
    
    // Low confidence warnings
    if (categoryResult.confidence < 0.5) {
      recommendations.push({
        type: 'warning',
        message: 'Category confidence is low. Please review and correct if necessary.'
      });
    }
    
    // Department assignment suggestions
    const departmentMap = {
      garbage_collection: 'Waste Management Department',
      road_maintenance: 'Public Works Department',
      street_lights: 'Electrical Department',
      water_supply: 'Water Supply Department',
      drainage: 'Drainage Department',
      public_safety: 'Police Department',
      health_services: 'Health Department',
      transportation: 'Transport Department'
    };
    
    if (categoryResult.category && departmentMap[categoryResult.category]) {
      recommendations.push({
        type: 'assignment',
        message: `Suggest assigning to ${departmentMap[categoryResult.category]}`
      });
    }
    
    // Priority escalation suggestions
    if (urgencyResult.category === 'emergency' || priorityResult.category === 'critical') {
      recommendations.push({
        type: 'escalation',
        message: 'Consider immediate escalation due to high priority/urgency'
      });
    }
    
    return recommendations;
  }
  
  /**
   * Estimate resolution time based on category and priority
   * @param {string} category - Report category
   * @param {string} priority - Report priority
   * @returns {number} - Estimated resolution time in hours
   */
  estimateResolutionTime(category, priority) {
    const baseTimes = {
      garbage_collection: 24,
      road_maintenance: 72,
      street_lights: 8,
      water_supply: 12,
      drainage: 48,
      public_safety: 2,
      health_services: 6,
      transportation: 48,
      noise_pollution: 24,
      air_pollution: 72,
      illegal_construction: 168,
      traffic_management: 24,
      public_facilities: 72,
      others: 48
    };
    
    const priorityMultipliers = {
      critical: 0.25,
      high: 0.5,
      medium: 1.0,
      low: 2.0
    };
    
    const baseTime = baseTimes[category] || 48;
    const multiplier = priorityMultipliers[priority] || 1.0;
    
    return Math.max(1, Math.round(baseTime * multiplier));
  }
  
  /**
   * Analyze report for duplicate detection
   * @param {Object} reportData - New report data
   * @param {Array} existingReports - Array of existing reports
   * @returns {Object} - Duplicate analysis result
   */
  async analyzeDuplicates(reportData, existingReports) {
    const duplicates = [];
    const { title, description, location, category } = reportData;
    
    const newReportText = `${title} ${description}`.toLowerCase();
    const newLocation = location.coordinates;
    
    for (const existingReport of existingReports) {
      let similarity = 0;
      
      // Text similarity (40% weight)
      const existingText = `${existingReport.title} ${existingReport.description}`.toLowerCase();
      const textSimilarity = this.calculateTextSimilarity(newReportText, existingText);
      similarity += textSimilarity * 0.4;
      
      // Location similarity (40% weight)
      if (existingReport.location?.coordinates && newLocation) {
        const distance = this.calculateDistance(
          newLocation.latitude, newLocation.longitude,
          existingReport.location.coordinates.latitude,
          existingReport.location.coordinates.longitude
        );
        
        // Consider reports within 1km as potentially similar
        const locationSimilarity = Math.max(0, 1 - (distance / 1));
        similarity += locationSimilarity * 0.4;
      }
      
      // Category similarity (20% weight)
      if (existingReport.category === category) {
        similarity += 0.2;
      }
      
      // If similarity is above threshold, consider it a potential duplicate
      if (similarity > 0.7) {
        duplicates.push({
          reportId: existingReport._id,
          similarity: similarity,
          factors: {
            textSimilarity: textSimilarity,
            locationDistance: this.calculateDistance(
              newLocation.latitude, newLocation.longitude,
              existingReport.location.coordinates.latitude,
              existingReport.location.coordinates.longitude
            ),
            sameCategory: existingReport.category === category
          }
        });
      }
    }
    
    return {
      isDuplicate: duplicates.length > 0,
      potentialDuplicates: duplicates,
      confidence: duplicates.length > 0 ? Math.max(...duplicates.map(d => d.similarity)) : 0
    };
  }
  
  /**
   * Calculate text similarity using simple word matching
   * @param {string} text1 - First text
   * @param {string} text2 - Second text
   * @returns {number} - Similarity score (0-1)
   */
  calculateTextSimilarity(text1, text2) {
    const words1 = new Set(text1.split(/\s+/));
    const words2 = new Set(text2.split(/\s+/));
    
    const intersection = new Set([...words1].filter(x => words2.has(x)));
    const union = new Set([...words1, ...words2]);
    
    return intersection.size / union.size;
  }
  
  /**
   * Calculate distance between two coordinates
   * @param {number} lat1 - Latitude 1
   * @param {number} lon1 - Longitude 1
   * @param {number} lat2 - Latitude 2
   * @param {number} lon2 - Longitude 2
   * @returns {number} - Distance in kilometers
   */
  calculateDistance(lat1, lon1, lat2, lon2) {
    const R = 6371; // Earth's radius in km
    const dLat = (lat2 - lat1) * Math.PI / 180;
    const dLon = (lon2 - lon1) * Math.PI / 180;
    const a = 
      Math.sin(dLat/2) * Math.sin(dLat/2) +
      Math.cos(lat1 * Math.PI / 180) * Math.cos(lat2 * Math.PI / 180) *
      Math.sin(dLon/2) * Math.sin(dLon/2);
    const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1-a));
    
    return R * c;
  }
}

module.exports = new SmartCategorizationService();